peps/pep-0740.rst

```diff
@@ -302,63 +322,47 @@ object) by selecting a new version number.
-      """
-
-The value of ``distribution`` is the same distribution filename that appears
-in the :pep:`503` and :pep:`691` APIs. For example, ``distribution`` would be
-``sampleproject-1.2.0-py2.py3-none-any.whl`` for the following simple index
-entry:
+Attestation statement and signature generation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-.. code-block:: html
+The *attestation statement* is the actual claim that is cryptographically signed
+over within the attestation object (i.e., the ``envelope.statement``).
 
-  <a href="https://example.com/...">sampleproject-1.2.0-py2.py3-none-any.whl</a><br/>
+The attestation statement is encoded as a
+`v1 in-toto Statement object <https://github.com/in-toto/attestation/blob/v1.0/spec/v1.0/statement.md>`__,
+in JSON form. When serialized the statement is treated as an opaque binary blob,
+avoiding the need for canonicalization. An example JSON-encoded statement is
+provided in :ref:`appendix-4`.
 
-In practice, this means that ``distribution`` is defined by the PyPA's
-living specifications for
-:ref:`binary distributions <packaging:binary-distribution-format>` and
-:ref:`source distributions <packaging:source-distribution-format>`, although
-non-conforming distributions may be hosted by the index.
+In addition to being a v1 in-toto Statement, the attestation statement is constrained
+in the following ways:
 
-The following pseudocode demonstrates the construction of an attestation
-payload and its signature:
+* The in-toto ``subject`` **MUST** contain only a single subject.
+* ``subject[0].name`` **MUST** be the distribution's filename, appropriately
+  normalized. See below for notes on sdist and wheel filename normalization.
+* ``subject[0].digest`` **MUST** contain a SHA-256 digest. Other digests
+  **MAY** be present. The digests **MUST** be represented as hexadecimal strings.
+* The following ``predicateType`` values are supported:
 
-.. code-block:: python
+  * `SLSA Provenance <https://slsa.dev/provenance/v1>`__: ``https://slsa.dev/provenance/v1``
+  * `PyPI Publish Attestation <https://docs.pypi.org/attestations/publish/v1>`__: ``https://docs.pypi.org/attestations/publish/v1``
 
-  def build_payload(dist: Path) -> AttestationPayload:
-      return AttestationPayload(
-        distribution=dist.name,
-        digest=sha256(dist.read_bytes()).hexdigest,
-      )
+For the distribution's filename, normalization should be performed according to
+the :ref:`binary distribution format <packaging:binary-distribution-format>` or
+:ref:`source distribution format <packaging:source-distribution-format>`, as
+appropriate. In particular, the distribution name **MUST** be lowercased and
+normalized, per :ref:`name normalization <packaging:name-normalization>`,
+followed by replacing runs of ``-`` with ``_``. Similar, the distribution version
+**MUST** be normalized per the
+:ref:`version specification <packaging:version-specifiers>`.
```

Comment:
I thought about this some more, and even this may not be sufficient, and will be annoying to do in practice:

1. `parse_wheel_filename` is a (somewhat) one-way operation, since it fully expands the tag set
2. The tag set itself isn't subject to any canonicalization rules, e.g. around the order of tags in each group in compressed tag sets

Given the above, it may just be easier to say that the distribution's filename needs to be a valid name, and is considered equal if it parses equally, meaning that the comparison is:

```python
# wheels
parse_wheel_filename(actual) == parse_wheel_filename(expected)

# sdists
parse_sdist_filename(actual) == parse_sdist_filename(expected)
```